Introduce Ansible roles and playbook for Ceph

fultonj opened the pull request:
These five roles and the ceph.yaml playbook can be used to deploy Ceph on EDPM nodes for testing HCI.

The cifmw_cephadm, cifmw_ceph_client, and cifmw_create_admin are from TripleO but only features needed for CI are included. The cifmw_block_device role is from tripleo-ci and creates a virtual block device, separate from the root volume, which can be used as an OSD. The cifmw_ceph_spec role is introduced here to render an opinionated Ceph spec for a CI job based on what is found on EDPM nodes. The cifmw_create_admin role can be used to create an SSH user for cephadm. Most of the work is from the cifmw_cephadm role, which is an Ansible wrapper to call the Ceph cephadm tool.

As a pull request owner and reviewers, I checked that:
- [x] Appropriate testing is done and actually running
- [X] Appropriate documentation exists and/or is up-to-date:
- [X] README in the role
- [X] Content of the docs/source is reflecting the changes

This is a big PR but most of it is copied from TripleO. Here's an overview to get started with reviewing:

The new playbook:

  https://github.com/fultonj/ci-framework/blob/cephadm/ci_framework/playbooks/ceph.yml

is described here:

  https://github.com/fultonj/ci-framework/blob/cephadm/ci_framework/roles/cifmw_cephadm/README.md

Its supporting Ansible roles in order of relevance are described here:

  https://github.com/fultonj/ci-framework/blob/cephadm/ci_framework/roles/cifmw_ceph_spec/README.md
  https://github.com/fultonj/ci-framework/blob/cephadm/ci_framework/roles/cifmw_ceph_client/README.md
  https://github.com/fultonj/ci-framework/blob/cephadm/ci_framework/roles/cifmw_block_device/README.md

raukadah commented on ci_framework/roles/cifmw_block_device/handlers/main.yml:
```diff
@@ -0,0 +1,15 @@
+---
```
`handlers` directory is empty. We can drop this it.

raukadah commented on ci_framework/roles/cifmw_block_device/vars/main.yml:
```diff
@@ -0,0 +1,22 @@
+---
```
`vars` directory is empty. We can drop this.

fultonj replied:
done

raukadah commented on ci_framework/roles/cifmw_ceph_spec/handlers/main.yml:
```diff
@@ -0,0 +1,15 @@
+---
```
`handlers` directory is emtpy. We can drop it.

fultonj replied:
done

raukadah commented on ci_framework/roles/cifmw_ceph_spec/tasks/cleanup.yml:
```diff
@@ -0,0 +1,19 @@
+---
```
There is no cleanup task here. We can also drop this file.

raukadah commented on ci_framework/roles/cifmw_ceph_spec/vars/main.yml:
```diff
@@ -0,0 +1,22 @@
+---
```
`vars` directory does not contain any vars currently. We can drop it.

fultonj replied:
done

raukadah commented on ci_framework/roles/cifmw_ceph_client/defaults/main.yml:
```diff
@@ -0,0 +1,35 @@
+cifmw_ceph_client_debug: "{{ (ansible_verbosity | int) >= 2 | bool }}"
+cifmw_ceph_client_hide_sensitive_logs: true
+cifmw_ceph_client_config_home: "/etc/ceph"
+tripleo_ceph_client_fetch_dir: "/tmp"
```
we can replace `tripleo_ceph_client_fetch_dir` to `cifmw_ceph_client_fetch_dir`.

fultonj replied:
done

raukadah commented on ci_framework/roles/cifmw_ceph_client/defaults/main.yml:
```diff
@@ -0,0 +1,35 @@
+cifmw_ceph_client_rbd_admin_socket_path: '/var/run/ceph'
+cifmw_ceph_client_rbd_log_path: '/var/log/ceph'
+cifmw_ceph_client_rbd_log_file: "{{ cifmw_ceph_client_rbd_log_path }}/qemu-guest-$pid.log"
+external_cluster_mon_ips: ''
```
We can replace `external_cluster_mon_ips` to `cifmw_ceph_client_external_cluster_mon_ips` . It will help to identify vars related to `cifmw_ceph_client` role.

raukadah commented on ci_framework/roles/cifmw_ceph_client/molecule/default/converge.yml:
```diff
@@ -0,0 +1,200 @@
+          register: cat_ceph_conf
+          when: verbosity
+        - debug:
+            msg: "{{ cat_ceph_conf.stdout }}"
```
We can also use `var: cat_ceph_conf.stdout` to print information from a particular variable. https://docs.ansible.com/ansible/latest/collections/ansible/builtin/debug_module.html#examples

fultonj replied:
done

raukadah commented on ci_framework/roles/cifmw_cephadm/tasks/apply_spec.yml:
```diff
@@ -0,0 +1,62 @@
+
+- name: Show results of spec apply
+  debug:
+    msg: "{{ cifmw_cephadm_apply_spec_out }}"
```
We can use `var: cifmw_cephadm_apply_spec_out` instead of msg.

fultonj replied:
done

raukadah reviewed (commented):
Thank you John for adding ceph roles to ci-framework and detailed usage in the commit message. I have done the initial review. It looks great.

There are some files with no relevant tasks and vars, I think we can drop it.
We can also drop `files/.gitkeep` files.

The molecule zuul jobs are missing. We can add them by running `make role_molecule` command. It will generate the relevant job definitions for the newly added ansible roles.

fultonj commented:
> Thank you John for adding ceph roles to ci-framework and detailed usage in the commit message. I have done the initial review. It looks great. I have also tested it locally by following instruction from [Tips on using standalone](https://github.com/fultonj/ci-framework/blob/cephadm/ci_framework/roles/cifmw_cephadm/README.md#tips-for-using-standalone).
> 
> Based on that I have two suggestions.
> 
> * We need to add ` ansible.netcommon` in https://github.com/openstack-k8s-operators/ci-framework/blob/main/requirements.yml and make sure netaddr is properly installed via pip.
> * We need to create a var file to store the value of `cifmw_cephadm_keys` (it's value is already defined in README) and pass that var file while calling ceph.yaml to avoid cifmw_cephadm_keys undefined error.
> 
> After applying these two suggestion, playbook executes successfully.
> 
> There are some files with no relevant tasks and vars, I think we can drop it. We can also drop `files/.gitkeep` files.
> 
> The molecule zuul jobs are missing. We can add them by running `make role_molecule` command. It will generate the relevant job definitions for the newly added ansible roles.

Thank you @raukadah for the review

I've made all of the changes except two. 

1. I will run `make role_molecule` for my next revision to the PR

2. Question about this comment "We need to create a var file to store the value of cifmw_cephadm_keys (it's value is already defined in README) and pass that var file while calling ceph.yaml to avoid cifmw_cephadm_keys undefined error."

I'm calling that role with the cifmw_cephadm_keys defined:

  https://github.com/fultonj/ci-framework/blob/cephadm/ci_framework/playbooks/ceph.yml#L165-L177

So I'm wondering how you got that error.

I could set a default in defaults/main.yml but I want to cephx key to be dynamically generated by the cephx_key module https://github.com/fultonj/ci-framework/blob/cephadm/ci_framework/playbooks/ceph.yml#L133-L135

fultonj commented:
> > Thank you John for adding ceph roles to ci-framework and detailed usage in the commit message. I have done the initial review. It looks great. I have also tested it locally by following instruction from [Tips on using standalone](https://github.com/fultonj/ci-framework/blob/cephadm/ci_framework/roles/cifmw_cephadm/README.md#tips-for-using-standalone).
> > Based on that I have two suggestions.
> > 
> > * We need to add ` ansible.netcommon` in https://github.com/openstack-k8s-operators/ci-framework/blob/main/requirements.yml and make sure netaddr is properly installed via pip.
> > * We need to create a var file to store the value of `cifmw_cephadm_keys` (it's value is already defined in README) and pass that var file while calling ceph.yaml to avoid cifmw_cephadm_keys undefined error.
> > 
> > After applying these two suggestion, playbook executes successfully.
> > There are some files with no relevant tasks and vars, I think we can drop it. We can also drop `files/.gitkeep` files.
> > The molecule zuul jobs are missing. We can add them by running `make role_molecule` command. It will generate the relevant job definitions for the newly added ansible roles.
> 
> Thank you @raukadah for the review
> 
> I've made all of the changes except two.
> 
> 1. I will run `make role_molecule` for my next revision to the PR
> 2. Question about this comment "We need to create a var file to store the value of cifmw_cephadm_keys (it's value is already defined in README) and pass that var file while calling ceph.yaml to avoid cifmw_cephadm_keys undefined error."
> 
> I'm calling that role with the cifmw_cephadm_keys defined:
> 
> https://github.com/fultonj/ci-framework/blob/cephadm/ci_framework/playbooks/ceph.yml#L165-L177
> 
> So I'm wondering how you got that error.
> 
> I could set a default in defaults/main.yml but I want to cephx key to be dynamically generated by the cephx_key module https://github.com/fultonj/ci-framework/blob/cephadm/ci_framework/playbooks/ceph.yml#L133-L135

As per our chat I understand the error you had was:
```
TASK [cifmw_cephadm : Save cifmw_ceph_client_vars file src=templates/ceph_client.yaml.j2, dest={{ cifmw_ceph_client_vars }}, mode=420, force=True] ***************************
An exception occurred during task execution. To see the full traceback, use -vvv. The error was: ansible.errors.AnsibleUndefinedVariable: 'cifmw_cephadm_keys' is undefined. 'cifmw_cephadm_keys' is undefined
fatal: [192.168.122.100 -> localhost]: FAILED! => {"changed": false, "msg": "AnsibleUndefinedVariable: 'cifmw_cephadm_keys' is undefined. 'cifmw_cephadm_keys' is undefined"}
```
I had mistakenly limited the scope of `cifmw_cephadm_keys` to only the `keys` task, but it is also needed for the `export` task. I've set `cifmw_cephadm_keys` as a fact so that won't happen again. Thanks for pointing this out.

https://github.com/openstack-k8s-operators/ci-framework/compare/ceccc731ad2b4b84f939e22e5c0ee0b5b86e544c..8bb2820566b6c556fcd610535f7e54c1af58dddc

fultonj commented:
@raukadah After I ran `pip install Jinja2==3.0.3` I no longer get the `TypeError: 'PosixPath' object is not iterable` error and `make role_molecule` completes without error. I'm now trying to get zuul to be happy with the zuul.d/projects.yaml file that `make role_molecule` produced however.

fultonj commented:
> @raukadah I'm now trying to get zuul to be happy with the zuul.d/projects.yaml file that `make role_molecule` produced however.

I was able to work around it by modifying the file manually.

fultonj commented:
> The molecule zuul jobs are missing.

They are now in place.

raukadah commented on zuul.d/molecule.yaml:
```diff
@@ -1,305 +1,354 @@
+    - ^ci_framework/roles/os_must_gather/(?!meta|README).*
+    - ^ci/playbooks/molecule.*
     name: cifmw-molecule-os_must_gather
-    parent: cifmw-molecule-base-crc
```
we can keep this parent `cifmw-molecule-base-crc` otherwise relevant molecule job will fail.

fultonj replied:
keeping

raukadah commented on zuul.d/molecule.yaml:
```diff
@@ -1,305 +1,354 @@
+    - ^ci_framework/roles/operator_deploy/(?!meta|README).*
+    - ^ci/playbooks/molecule.*
     name: cifmw-molecule-operator_deploy
-    nodeset: centos-9-crc-xl
```
We need to keep this nodeset otherwise molecule job will fail.

fultonj replied:
keeping

raukadah commented on zuul.d/molecule.yaml:
```diff
@@ -1,305 +1,354 @@
+    - ^ci_framework/roles/set_openstack_containers/(?!meta|README).*
+    - ^ci/playbooks/molecule.*
     name: cifmw-molecule-set_openstack_containers
-    parent: cifmw-molecule-base-crc
```
We need to keep this parent to avoid to molecule job failure.

fultonj replied:
keeping

raukadah reviewed (commented):
Thank you @fultonj for addressing the suggestions and adding molecule tests. It looks good. 

I think `make role_molecule` command messed some of the job definition( sorry for that, I will fix that in a new patch).  Till now, I have suggested to add back few jobs nodeset and parent. It will fix the molecule job.  It will help to get this large pr in.

fultonj commented on ci_framework/roles/cifmw_block_device/tasks/main.yml:
```diff
@@ -0,0 +1,64 @@
+    force: true
+
+- name: Ensure ceph-osd-losetup is running and enabled
+  tags: systemd
```
done

fultonj commented:
> Thank you @fultonj for addressing the suggestions and adding molecule tests. It looks good.
> 
> I think `make role_molecule` command messed some of the job definition( sorry for that, I will fix that in a new patch). Till now, I have suggested to add back few jobs nodeset and parent. It will fix the molecule job. It will help to get this large pr in.

Thanks @raukadah I've made your most recent request for changes. I'm still fixing molecule tests in my env (got a lot of pings for other things today)

https://github.com/openstack-k8s-operators/ci-framework/compare/8e4db9048caa6fe6091ad59040d854f2b0280b5e..f87cb6490c379183157b1437da9ceca062ec91f9

raukadah commented:
> > Thank you @fultonj for addressing the suggestions and adding molecule tests. It looks good.
> > I think `make role_molecule` command messed some of the job definition( sorry for that, I will fix that in a new patch). Till now, I have suggested to add back few jobs nodeset and parent. It will fix the molecule job. It will help to get this large pr in.
> 
> Thanks @raukadah I've made your most recent request for changes. I'm still fixing molecule tests in my env (got a lot of pings for other things today)
> 
> https://github.com/openstack-k8s-operators/ci-framework/compare/8e4db9048caa6fe6091ad59040d854f2b0280b5e..f87cb6490c379183157b1437da9ceca062ec91f9

Thank you @fultonj for addressing all the issue. https://github.com/openstack-k8s-operators/ci-framework/pull/410 will fix the molecule failure.

raukadah commented:
recheck

fultonj commented:
> > > Thank you @fultonj for addressing the suggestions and adding molecule tests. It looks good.
> > > I think `make role_molecule` command messed some of the job definition( sorry for that, I will fix that in a new patch). Till now, I have suggested to add back few jobs nodeset and parent. It will fix the molecule job. It will help to get this large pr in.
> > 
> > 
> > Thanks @raukadah I've made your most recent request for changes. I'm still fixing molecule tests in my env (got a lot of pings for other things today)
> > https://github.com/openstack-k8s-operators/ci-framework/compare/8e4db9048caa6fe6091ad59040d854f2b0280b5e..f87cb6490c379183157b1437da9ceca062ec91f9
> 
> Thank you @fultonj for addressing all the issue. #410 will fix the molecule failure.

Yes, it did. Thanks @raukadah . Molecule tests should now be green.

raukadah commented on zuul.d/projects.yaml:
```diff
@@ -12,34 +12,40 @@
+      - cifmw-molecule-set_openstack_containers
+      - cifmw-molecule-tempest
+      - cifmw-molecule-test_deps
+      - cifmw-molecule-build_containers
```
Can we drop this job from here? this pr https://github.com/openstack-k8s-operators/ci-framework/pull/285 includes this job with fixes. It will help to merge this pr.

fultonj replied:
I think this is done. Here's the diff. Please let me know if I misunderstood. 

https://github.com/openstack-k8s-operators/ci-framework/compare/d85be4e1acbf64f76f44f511316cc12ce577e48c..46f76cd4b59cacd395c2c21d1ede4b25ea9951a4

raukadah commented on zuul.d/molecule.yaml:
```diff
@@ -246,60 +307,59 @@
     parent: cifmw-molecule-base
     vars:
-      TEST_RUN: ci_setup
+      TEST_RUN: build_containers
```
We can also drop the job definition.

raukadah reviewed (commented):
Thank you John for addressing all the comments. We are almost ready to merge.

We can drop the job definition of `cifmw-molecule-build_containers` from this pr. Then this patch is good to merge.

raukadah commented:
/override build

raukadah reviewed (approved):
Thank you @fultonj , It is a great addition to our ci-framework role collection :-)

viroel commented:
great job @fultonj - thanks for adding this.
Some initial tests with external ceph, deployed using ceph.yaml playbook:
Glance image:

> cephadm shell -- rbd -p images ls -l
> NAME                                       SIZE    PARENT  FMT  PROT  LOCK
> 450e2cdd-ba09-4e6e-932c-291f6145c423       16 MiB            2            
> 450e2cdd-ba09-4e6e-932c-291f6145c423@snap  16 MiB            2  yes

cinder backend:

> $ openstack volume backend pool list
> +-----------------------------------+
> | Name                              |
> +-----------------------------------+
> | cinder-volume-volume1-0@ceph#ceph |
> +-----------------------------------+
